commandapi-platforms/commandapi-paper/commandapi-paper-core/src/main/java/dev/jorel/commandapi/PaperCommandRegistration.java

```diff
@@ -0,0 +1,155 @@
+	private final Supplier<CommandDispatcher<Source>> getBrigadierDispatcher;
+	private final Runnable reloadHelpTopics;
+	private final Predicate<CommandNode<Source>> isBukkitCommand;
+
+	private final boolean[] lifecycleEventRegistered = new boolean[2];
+	private final CommandDispatcher<CommandSourceStack> bootstrapDispatcher = new CommandDispatcher<>();
+	private final CommandDispatcher<CommandSourceStack> pluginDispatcher = new CommandDispatcher<>();
+
+	public PaperCommandRegistration(Supplier<CommandDispatcher<Source>> getBrigadierDispatcher, Runnable reloadHelpTopics, Predicate<CommandNode<Source>> isBukkitCommand) {
+		this.getBrigadierDispatcher = getBrigadierDispatcher;
+		this.reloadHelpTopics = reloadHelpTopics;
+		this.isBukkitCommand = isBukkitCommand;
+	}
+
+	// Provide access to internal functions that may be useful to developers
+
+	/**
+	 * Checks if a Brigadier command node came from wrapping a Bukkit command
+	 *
+	 * @param node The CommandNode to check
+	 * @return true if the CommandNode is being handled by Paper's BukkitCommandNode
+	 */
+	public boolean isBukkitCommand(CommandNode<Source> node) {
+		return isBukkitCommand.test(node);
+	}
+
+	public CommandDispatcher<Source> getPaperDispatcher() {
+		return CommandAPIPaper.getPaper().getNMS().getPaperCommandDispatcher();
+	}
+
+	// Implement CommandRegistrationStrategy methods
+	@Override
+	public CommandDispatcher<Source> getBrigadierDispatcher() {
+		return getBrigadierDispatcher.get();
+	}
+
+	@Override
+	public void runTasksAfterServerStart() {
+		// Nothing to do
+	}
+
+	@Override
+	public void postCommandRegistration(RegisteredCommand registeredCommand, LiteralCommandNode<Source> resultantNode, List<LiteralCommandNode<Source>> aliasNodes) {
+		// Nothing to do
+	}
+
+	@Override
+	public LiteralCommandNode<Source> registerCommandNode(LiteralArgumentBuilder<Source> node, String namespace) {
+		LiteralCommandNode<Source> built = node.build();
+		if (Bukkit.getServer() == null) {
+			bootstrapDispatcher.getRoot().addChild((CommandNode<CommandSourceStack>) built);
+		} else {
+			pluginDispatcher.getRoot().addChild((CommandNode<CommandSourceStack>) built);
+		}
+		return built;
+	}
+
+	@Override
+	public void unregister(String commandName, boolean unregisterNamespaces, boolean unregisterBukkit) {
+		// Remove nodes from the dispatcher
+		removeBrigadierCommands(getBrigadierDispatcher().getRoot(), commandName, unregisterNamespaces,
+			// If we are unregistering a Bukkit command, ONLY unregister BukkitCommandNodes
+			// If we are unregistering a Vanilla command, DO NOT unregister BukkitCommandNodes
+			c -> !unregisterBukkit ^ isBukkitCommand.test(c));
+
+		// Update the dispatcher file
+		CommandAPIHandler.getInstance().writeDispatcherToFile();
+	}
+
+	@Override
+	public void preReloadDataPacks() {
+		reloadHelpTopics.run();
+		CommandAPIBukkit.get().updateHelpForCommands(CommandAPI.getRegisteredCommands());
+	}
+
+	void registerLifecycleEvent(boolean bootstrap) {
```

Comment:
When this method is called, the argument is always `Bukkit.getServer() == null`. It seems like this could be moved into this method.

```suggestion
	void registerLifecycleEvent() {
		boolean bootstrap = Bukkit.getServer() == null;
```